File: journal/consensus/poet/poet_enclave_simulator/wait_timer.cpp
// Copyright 2016 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------------

#ifdef _WIN32
    #include <windows.h>
    #include <random>
#else
    #include <sys/time.h>
#endif
#ifdef __APPLE__
    #include <random>
#endif
#include <algorithm>
#include <memory>
#include <time.h>
#include <stdlib.h>
#include <string>
#include <json-c/json.h>
#include <iostream>

#include "poet_enclave.h"
#include "common.h"

using namespace std;

static double ComputeDuration(double mean)
{
#ifdef _WIN32
    SYSTEMTIME seed;GetLocalTime(&seed);
    std::tr1::mt19937 generator((int) seed.wMilliseconds);
    std::tr1::exponential_distribution<double> distribution(1.0 / mean);
#else
    struct timeval seed;
    gettimeofday(&seed, NULL);

    std::default_random_engine generator(seed.tv_usec);
    std::exponential_distribution<double> distribution(1.0 / mean);
#endif

    double wait = distribution(generator);

    return (MINIMUM_WAIT_TIME + wait);
}

WaitTimer::WaitTimer(string encoded, string signature)
{
    this->signature = signature;
    this->deserialize(encoded);
}

WaitTimer::WaitTimer(string previous_certificate_id, double local_mean)
{
    if(local_mean <= 0) {
        throw ValueError("Invalid local mean");
    }

    if(previous_certificate_id.length() != IDENTIFIER_LENGTH) {
        throw ValueError("Invalid previous_certificate_id");
    }

    this->local_mean = local_mean;

    this->request_time = CurrentTime();
    this->duration = ComputeDuration(local_mean);

    this->previous_certificate_id = previous_certificate_id;
}

bool WaitTimer::is_expired(void)
{
    return (request_time + duration) < CurrentTime() ? true : false;
}

bool WaitTimer::deserialize(string serialized)
{
    json_object *jobj = json_tokener_parse((char *)serialized.data());
    if (jobj == NULL)
    {
        return false;
    }

    struct json_object *obj = NULL;

    // Use alphabetical order for the keys
    if (json_object_object_get_ex(jobj, "Duration", &obj))
        duration = json_object_get_double(obj);

    if (json_object_object_get_ex(jobj, "LocalMean", &obj))
        local_mean = json_object_get_double(obj);

    if (json_object_object_get_ex(jobj, "PreviousCertID", &obj))
        previous_certificate_id = json_object_get_string(obj);

    if (json_object_object_get_ex(jobj, "RequestTime", &obj))
        request_time = json_object_get_double(obj);

    return true;
}

string WaitTimer::serialize()
{
    json_object *jobj = json_object_new_object();

    // Use alphabetical order for the keys
    json_object_object_add(jobj, "Duration", json_object_new_double(duration));
    json_object_object_add(jobj, "LocalMean", json_object_new_double(local_mean));
    json_object_object_add(jobj, "PreviousCertID", json_object_new_string((char *)previous_certificate_id.data()));
    json_object_object_add(jobj, "RequestTime", json_object_new_double(request_time));

    string serialized = (char *)json_object_to_json_string(jobj);

    return serialized;
}


WaitTimer* create_wait_timer(string prev_cert_id,
                                double local_mean)
{
    WaitTimer *timer = new WaitTimer(prev_cert_id, local_mean);
    timer->signature = SignMessage(WaitTimerPrivateKey, timer->serialize());
    return(timer);
}

WaitTimer* deserialize_wait_timer(string serialized_timer,
                                        string signature)
{
    if(!verify_signature(WaitTimerPublicKey, serialized_timer, signature))
    {
        throw ValueError("Signature failed to verify.");
    }

    WaitTimer* timer = new WaitTimer(serialized_timer, signature);

    return timer;
}

File: consensus/poet/core/sawtooth_poet/poet_consensus/wait_timer.py
# Copyright 2016, 2017 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------------

import logging

LOGGER = logging.getLogger(__name__)


class WaitTimer(object):
    """Wait timers represent a random duration incorporated into a wait
    certificate.

    Attributes:
        WaitTimer.minimum_wait_time (float): The minimum wait time in seconds.
        WaitTimer.target_wait_time (float): The target wait time in seconds.
        WaitTimer.initial_wait_time (float): The initial wait time in seconds.
        WaitTimer.certificate_sample_length (int): The number of certificates
            to sample for the population estimate.
        WaitTimer.fixed_duration_blocks (int): If fewer than
            WaitTimer.fixed_duration_blocks exist, then base the local mean
            on a ratio based on InitialWaitTime, rather than the history.
        previous_certificate_id (str): The id of the previous certificate.
        local_mean (float): The local mean wait time based on the history of
            certs.
        request_time (float): The request time.
        duration (float): The duration of the wait timer.
        validator_address (str): The address of the validator that created the
            wait timer
    """
    minimum_wait_time = 1.0
    target_wait_time = 20.0
    initial_wait_time = 3000.0
    certificate_sample_length = 50
    fixed_duration_blocks = certificate_sample_length

    @classmethod
    def create_wait_timer(cls,
                          poet_enclave_module,
                          validator_address,
                          previous_certificate_id,
                          consensus_state,
                          poet_config_view):
        """Creates a wait timer in the enclave and then constructs
        a WaitTimer object.

        Args:
            poet_enclave_module (module): The module that implements the
                underlying PoET enclave.
            validator_address (str): A string representing the address of the
                validator creating the wait timer.
            previous_certificate_id (str): The ID of the wait certificate for
                the block attempting to build upon
            consensus_state (ConsensusState): The current PoET consensus state
            poet_config_view (PoetConfigView): The current PoET config view

        Returns:
            journal.consensus.poet.wait_timer.WaitTimer: A new wait timer.
        """

        # Create an enclave timer object and then use it to create a
        # WaitTimer object
        enclave_timer = \
            poet_enclave_module.create_wait_timer(
                validator_address,
                previous_certificate_id,
                consensus_state.compute_local_mean(
                    poet_config_view=poet_config_view),
                poet_config_view.minimum_wait_time)

        return cls(enclave_timer)

    @property
    def population_estimate(self):
        return self.local_mean / WaitTimer.target_wait_time

    @property
    def enclave_wait_timer(self):
        """Converts the serialized timer into an eclave timer object.

        Returns:
            <poet_enclave_module>.WaitTimer: The deserialized enclave timer
                object.
        """
        return self._enclave_wait_timer

    def __init__(self, enclave_timer):
        self.previous_certificate_id =\
            str(enclave_timer.previous_certificate_id)
        self.local_mean = float(enclave_timer.local_mean)
        self.request_time = float(enclave_timer.request_time)
        self.duration = float(enclave_timer.duration)
        self.validator_address = str(enclave_timer.validator_address)

        self._enclave_wait_timer = enclave_timer
        self._expires = self.request_time + self.duration + 0.1
        self._serialized_timer = None

    def __str__(self):
        return \
            'TIMER, {0:0.2f}, {1:0.2f}, {2}'.format(
                self.local_mean,
                self.duration,
                self.previous_certificate_id)

    def serialize(self):
        """Serializes the underlying enclave wait timer
        """
        if self._serialized_timer is None:
            self._serialized_timer = self._enclave_wait_timer.serialize()

        return self._serialized_timer

    def has_expired(self, now):
        """Determines whether the timer has expired.

        Args:
            now (float): The current time.

        Returns:
            bool: True if the timer has expired, false otherwise.
        """
        if now < self._expires:
            return False

        return self._enclave_wait_timer.has_expired()


def set_wait_timer_globals(target_wait_time=None,
                           initial_wait_time=None,
                           certificate_sample_length=None,
                           fixed_duration_blocks=None,
                           minimum_wait_time=None):
    if target_wait_time is not None:
        WaitTimer.target_wait_time = float(target_wait_time)
    if initial_wait_time is not None:
        WaitTimer.initial_wait_time = float(initial_wait_time)
    if certificate_sample_length is not None:
        WaitTimer.certificate_sample_length = int(certificate_sample_length)
        WaitTimer.fixed_duration_blocks = int(certificate_sample_length)
    if fixed_duration_blocks is not None:
        WaitTimer.fixed_duration_blocks = int(fixed_duration_blocks)
    if minimum_wait_time is not None:
        WaitTimer.minimum_wait_time = float(minimum_wait_time)
